File: src/main/java/bad/robot/radiate/teamcity/Authorisation.java
package bad.robot.radiate.teamcity;

enum Authorisation {

    GuestAuthorisation("guestAuth"),
    BasicAuthorisation("httpAuth");

    private final String pathSegment;

    Authorisation(String pathSegment) {
        this.pathSegment = pathSegment;
    }

    public String toString() {
        return pathSegment;
    }

    public static Authorisation authorisationFor(Username username, Password password) {
        if (username.equals(new NoUsername()) || password.equals(new NoPassword()))
            return GuestAuthorisation;
        return BasicAuthorisation;
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/Authorisation.scala
package bad.robot.radiate.teamcity

import bad.robot.radiate.config.{NoPassword, Password}

sealed trait Authorisation {
  val pathSegment: String
  override def toString = pathSegment
}

object Authorisation {
  def authorisationFor(username: Username, password: Password): Authorisation = {
    if ((username == NoUsername) || (password == NoPassword))
      GuestAuthorisationS
    else
      BasicAuthorisationS
  }
}

case object GuestAuthorisationS extends Authorisation {
  override val pathSegment = "guestAuth"
}

case object BasicAuthorisationS extends Authorisation {
  override val pathSegment = "httpAuth"
}